third_party/intel/include/TritonGENToLLVM/Attributes.h

```diff
@@ -0,0 +1,216 @@
+
+  AttributeList() = default;
+  AttributeList(const mlir::NamedAttrList &FnAttrs,
+                const mlir::NamedAttrList &RetAttrs,
+                llvm::ArrayRef<mlir::NamedAttrList> ParamAttrs);
+
+  //===--------------------------------------------------------------------===//
+  // AttributeList Mutation
+  //===--------------------------------------------------------------------===//
+
+  /// Add function. return value, and parameters attributes to the list.
+  AttributeList &addAttrs(const AttrBuilder &FnAttrB,
+                          const AttrBuilder &RetAttrB,
+                          llvm::ArrayRef<mlir::NamedAttrList> Attrs);
+
+  /// Add function attributes to the list.
+  AttributeList &addFnAttrs(const AttrBuilder &B);
+  AttributeList &addFnAttrs(const mlir::NamedAttrList &Attrs,
+                            mlir::MLIRContext &Ctx);
+
+  /// Add return value attributes to the list.
+  AttributeList &addRetAttrs(const AttrBuilder &B);
+  AttributeList &addRetAttrs(const mlir::NamedAttrList &Attrs,
+                             mlir::MLIRContext &Ctx);
+
+  /// Add parameters attributes to the list.
+  AttributeList &addParamAttributes(llvm::ArrayRef<mlir::NamedAttrList> Attrs);
+
+  /// The function attributes are returned.
+  const mlir::NamedAttrList &getFnAttributes() const { return FnAttrs; }
+
+  /// The attributes for the ret value are returned.
+  const mlir::NamedAttrList &getRetAttributes() const { return RetAttrs; }
+
+  /// The attributes for the parameters are returned.
+  const mlir::ArrayRef<mlir::NamedAttrList> getParamAttributes() const {
+    return ParamAttrs;
+  }
+
+private:
+  /// The attributes that we are managing.
+  mlir::NamedAttrList FnAttrs;
+  mlir::NamedAttrList RetAttrs;
+  llvm::SmallVector<mlir::NamedAttrList, 8> ParamAttrs;
+};
+
+/// \class
+/// Facilitates the construction of LLVM dialect attributes for a particular
+/// argument, parameter, function, or return value.
+class AttrBuilder {
+public:
+  AttrBuilder(mlir::MLIRContext &Ctx) : Ctx(Ctx) {}
+  AttrBuilder(const AttrBuilder &) = delete;
+  AttrBuilder(AttrBuilder &&) = default;
+
+  /// Add the LLVM attribute identified by \p Kind to the builder.
+  AttrBuilder &addAttribute(llvm::Attribute::AttrKind Kind);
+
+  /// Add the LLVM attribute identified by \p Kind with a type given by \p Ty
+  /// to the builder.
+  AttrBuilder &addAttribute(llvm::Attribute::AttrKind Kind, mlir::Type Ty);
+
+  /// Add the LLVM attribute identified by \p Kind with a value given by \p Val
+  /// to the builder.
+  AttrBuilder &addAttribute(llvm::Attribute::AttrKind Kind, uint64_t Val);
+
+  /// Create a NamedAttribute with name \p AttrName and value \p Attr and add it
+  /// to the builder.
+  AttrBuilder &addAttribute(llvm::Twine AttrName, mlir::Attribute Attr);
+
+  /// Add the LLVM attribute identified by \p Kind to the builder "passthrough"
+  /// named attribute.
+  AttrBuilder &addPassThroughAttribute(llvm::Attribute::AttrKind Kind);
```

Comment:
passthrough is a word.
```suggestion
  AttrBuilder &addPassthroughAttribute(llvm::Attribute::AttrKind Kind);
```

Reply:
Correct.